From 879b89efe24c03bbd8d1f34b7c82b0ff65a85516 Mon Sep 17 00:00:00 2001
From: "qi.duan" <qi.duan@amlogic.com>
Date: Fri, 6 Dec 2013 23:09:53 +0800
Subject: [PATCH 2460/5965] PD #83999 new reserve memory way

Conflicts:

	arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
	arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd
	arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd
---
 arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd  |  87 ++++++++----
 arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd  |  48 +++----
 arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd  |  51 +++----
 arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd  |  44 +++---
 arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd  |  44 +++---
 arch/arm/boot/dts/amlogic/meson8_skt.dtd      |  48 +++----
 arch/arm/kernel/devtree.c                     |   4 +
 drivers/amlogic/amports/amstream.c            |  30 +++-
 drivers/amlogic/amports/encoder.c             |  16 ++-
 drivers/amlogic/deinterlace/deinterlace.c     |  14 +-
 .../aml_logo/logo_display/logo_output.c       |  17 ++-
 drivers/amlogic/display/osd/osd_main.c        |  14 ++
 drivers/amlogic/display/osd_ext/osd_main.c    |  15 +-
 drivers/amlogic/ion_dev/dev_ion.c             |  14 ++
 drivers/amlogic/ppmgr/ppmgr_drv.c             |  17 ++-
 drivers/amlogic/tvin/vdin/vdin_drv.c          |  13 ++
 drivers/amlogic/video_dev/amlvideo2.c         |  15 +-
 drivers/of/fdt.c                              | 131 +++++++++++++++++-
 include/linux/amlogic/amports/dsp_register.h  |   4 +-
 include/linux/of_fdt.h                        |   8 ++
 20 files changed, 467 insertions(+), 167 deletions(-)

diff --git a/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
index dcb128e81f65..a912eb91a209 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
@@ -76,9 +76,9 @@ void root_func(){
 //$$ L2 PROP_U32 6 = "linux,usable-memory"
 	memory{
 		device_type = "memory";
-		aml_reserved_start = <0x05000000>; /**reserved memory start */
-		aml_reserved_end = <0x15efffff>;/**reserved memory end */
-		linux,usable-memory = <0x00000000 0x04000000 0x16f00000 0x29100000>;//0x12f00000+0x2d100000=0x40000000 1G ddr
+		aml_reserved_start = <0x06000000>; /**reserved memory start */
+		aml_reserved_end = <0x05000000>;/**reserved memory end */
+		linux,total-memory = <0x40000000>;
 	};
 /// ***************************************************************************************
 ///	-	GIC
@@ -109,17 +109,6 @@ void root_func(){
 		    	6: 318.75M		7: 364.3M					 */
 	};
 	
-/// ***************************************************************************************
-///	-	ION
-//$$ MODULE="ION"
-//$$ DEVICE="ion_dev"
-//$$ L2 PROP_STR = "status"
-    ion_dev{
-        compatible = "amlogic,ion_dev";
-        dev_name = "ion_dev";
-        status = "ok";
-        reg = <0xb400000 0x01400000>;//same as following ppmgr
-    };
 /// **************************************************************************************	
 /// -   DISP&MM-FB
 //$$ MODULE = "DISP&MM-FB"	
@@ -132,9 +121,7 @@ void root_func(){
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reg = <0x05100000 0x01800000
-	   	 		0x06900000 0x00100000>;
-		rotation = <0>;
+	   	reserve-memory = <0x01800000  0x00100000>;
         vmode = <1>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
  		display_size_default = <2048 1536 2048 3072 32>;//2048x1536x8=0x01800000
 	};
@@ -149,8 +136,7 @@ void root_func(){
 		compatible = "amlogic,mesonstream";
 		dev_name = "mesonstream.0";
 		status = "okay";
-		reg = <0x06a00000 0x04000000
-			0xaa00000 0x00a00000>;
+		reserve-memory = <0x04000000 0x00a00000>;
 	};
 
 
@@ -164,7 +150,7 @@ void root_func(){
 		compatible = "amlogic,ppmgr";
 		dev_name = "ppmgr";
 		status = "okay";
-		reg = <	0xb400000 0x01400000>;
+		reserve-memory = <0x01400000>;
 	};
 
 /// ***************************************************************************************
@@ -177,7 +163,7 @@ void root_func(){
 		compatible = "amlogic,amvenc_avc";
 		dev_name = "amvenc_avc.0";
 		status = "okay";
-		reg = <0xc800000 0x00f00000>;
+		reserve-memory = <0x00f00000>;
 	};
 
 /// ***************************************************************************************
@@ -192,7 +178,7 @@ void root_func(){
         compatible = "amlogic,vdin";
         dev_name = "vdin0";
         status = "ok";
-	        reg = <	0xd700000 0x04000000>;
+	 reserve-memory = <0x04000000>;
         irq = <115>;
         vdin_id = <0>;
 	};
@@ -206,7 +192,7 @@ void root_func(){
         compatible = "amlogic,vdin";
         dev_name = "vdin1";
         status = "ok";
-		reg = <0x11700000 0x04000000>;
+	 reserve-memory = <0x04000000>;
         irq = <117>;
         vdin_id = <1>;
 	};
@@ -221,10 +207,21 @@ void root_func(){
 		compatible = "amlogic,amlvideo2";
 		dev_name = "amlvideo2.0";
 		status = "okay";
-		reg = <0x15700000 0x01800000>;
+		reserve-memory = <0x01800000>;
 	};
 
 /// ***************************************************************************************
+///	-	ION
+//$$ MODULE="ION"
+//$$ DEVICE="ion_dev"
+//$$ L2 PROP_STR = "status"
+    ion_dev{
+        compatible = "amlogic,ion_dev";
+        dev_name = "ion_dev";
+        status = "ok";
+        reserve-memory = <0x01400000>;
+    };
+/// ***************************************************************************************
 ///	-	DISP&MM-Vout
 //$$ MODULE = "DISP&MM-Vout"
 //$$ DEVICE = "mesonvout"
@@ -1436,6 +1433,48 @@ sdio{
 			bt_path = "gpio";
 			status = "okay";
 		};
+ ///	-	hm2057
+//$$ DEVICE="hm2057"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 = "front_back"
+//$$ L2 PROP_STR = "i2c_bus"
+//$$ L2 PROP_STR = "gpio_pwdn"
+//$$ L2 PROP_STR = "gpio_rst"
+//$$ L2 PROP_U32 = "mirror_flip"
+//$$ L2 PROP_U32 = "vertical_flip"
+//$$ L2 PROP_STR = "bt_path"    
+            	 cam_7{
+			cam_name = "hm2057";
+			front_back = <0>;
+			i2c_bus = "i2c_bus_d";
+			gpio_pwdn = "GPIOH_5";
+			gpio_rst = "GPIOH_4";
+			mirror_flip = <0>;
+			vertical_flip = <0>;	
+			bt_path = "gpio";
+			status = "okay";
+		};
+
+//$$ DEVICE="sp0a19"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 = "front_back"
+//$$ L2 PROP_STR = "i2c_bus"
+//$$ L2 PROP_STR = "gpio_pwdn"
+//$$ L2 PROP_STR = "gpio_rst"
+//$$ L2 PROP_U32 = "mirror_flip"
+//$$ L2 PROP_U32 = "vertical_flip"
+//$$ L2 PROP_STR = "bt_path"
+        cam_8{
+			cam_name = "sp0a19";
+			front_back = <0>;
+			i2c_bus = "i2c_bus_d";
+			gpio_pwdn = "GPIOH_6";
+			gpio_rst = "GPIOH_4";
+			mirror_flip = <0>;
+			vertical_flip = <0>;	
+			bt_path = "gpio";
+			status = "okay";
+		};
 	};	
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd
index a6f36112efb0..76f31e06e99c 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k100_v1.dtd
@@ -76,9 +76,9 @@ void root_func(){
 //$$ L2 PROP_U32 6 = "linux,usable-memory"
 	memory{
 		device_type = "memory";
-		aml_reserved_start = <0x05000000>; /**reserved memory start */
-		aml_reserved_end = <0x25dfffff>;/**reserved memory end */
-		linux,usable-memory = <0x00000000 0x04000000 0x25e00000 0x5a200000>;//0x25f00000+0x5a100000=0x80000000 2GB DDR
+		aml_reserved_start = <0x06000000>; /**reserved memory start */
+		aml_reserved_end = <0x05000000>;/**reserved memory end */
+		linux,total-memory = <0x80000000>;
 	};
 /// ***************************************************************************************
 ///	-	GIC
@@ -109,17 +109,6 @@ void root_func(){
 		    	6: 318.75M		7: 364.3M					 */
 	};
 	
-/// ***************************************************************************************
-///	-	ION
-//$$ MODULE="ION"
-//$$ DEVICE="ion_dev"
-//$$ L2 PROP_STR = "status"
-    ion_dev{
-        compatible = "amlogic,ion_dev";
-        dev_name = "ion_dev";
-        status = "ok";
-        reg = <0x1a300000 0x01400000>;//same as following ppmgr
-    };
 /// **************************************************************************************	
 /// -   DISP&MM-FB
 //$$ MODULE = "DISP&MM-FB"	
@@ -132,9 +121,7 @@ void root_func(){
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reg = <0x05100000 0x01800000
-	   	 		0x06900000 0x00100000>;
-		rotation = <0>;
+	   	reserve-memory = <0x01800000  0x00100000>;
         vmode = <1>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
  		display_size_default = <2048 1536 2048 3072 32>;//2048x1536x8=0x01800000
 	};
@@ -146,8 +133,7 @@ void root_func(){
 		compatible = "amlogic,mesonfb-ext";
 		dev_name = "mesonfb-ext";
 		status = "ok";
-		reg = <0x06a00000 0x01800000
-	       	 0x08200000 0x00100000>;
+		reserve-memory = <0x01800000  0x00100000>; 	 
 		display_size_default = <2048 1536 2048 3072 32>;//2048x1536x8=0x01800000
 	 
 	};
@@ -162,8 +148,7 @@ void root_func(){
 		compatible = "amlogic,mesonstream";
 		dev_name = "mesonstream.0";
 		status = "okay";
-		reg = <0x08300000 0x10000000
-			0x18300000 0x02000000>;
+		reserve-memory = <0x10000000 0x02000000>;
 	};
 
 
@@ -177,7 +162,7 @@ void root_func(){
 		compatible = "amlogic,ppmgr";
 		dev_name = "ppmgr";
 		status = "okay";
-		reg = <	0x1a300000 0x01400000>;
+		reserve-memory = <0x01400000>;
 	};
 
 /// ***************************************************************************************
@@ -190,7 +175,7 @@ void root_func(){
 		compatible = "amlogic,amvenc_avc";
 		dev_name = "amvenc_avc.0";
 		status = "okay";
-		reg = <0x1b700000 0x00f00000>;
+		reserve-memory = <0x00f00000>;
 	};
 
 /// ***************************************************************************************
@@ -205,7 +190,7 @@ void root_func(){
         compatible = "amlogic,vdin";
         dev_name = "vdin0";
         status = "ok";
-	        reg = <	0x1c600000 0x04000000>;
+	 reserve-memory = <0x04000000>;
         irq = <115>;
         vdin_id = <0>;
 	};
@@ -219,7 +204,7 @@ void root_func(){
         compatible = "amlogic,vdin";
         dev_name = "vdin1";
         status = "ok";
-		reg = <0x20600000 0x04000000>;
+	 reserve-memory = <0x04000000>;
         irq = <117>;
         vdin_id = <1>;
 	};
@@ -234,10 +219,21 @@ void root_func(){
 		compatible = "amlogic,amlvideo2";
 		dev_name = "amlvideo2.0";
 		status = "okay";
-		reg = <0x24600000 0x01800000>;
+		reserve-memory = <0x01800000>;
 	};
 
 /// ***************************************************************************************
+///	-	ION
+//$$ MODULE="ION"
+//$$ DEVICE="ion_dev"
+//$$ L2 PROP_STR = "status"
+    ion_dev{
+        compatible = "amlogic,ion_dev";
+        dev_name = "ion_dev";
+        status = "ok";
+        reserve-memory = <0x01400000>;
+    };
+/// ***************************************************************************************
 ///	-	DISP&MM-Vout
 //$$ MODULE = "DISP&MM-Vout"
 //$$ DEVICE = "mesonvout"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd
index 047ee93f5d0f..6ac6ca4902ad 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k101_v1.dtd
@@ -76,9 +76,9 @@ void root_func(){
 //$$ L2 PROP_U32 6 = "linux,usable-memory"
 	memory{
 		device_type = "memory";
-		aml_reserved_start = <0x05000000>; /**reserved memory start */
-		aml_reserved_end = <0x173fffff>;/**reserved memory end */
-		linux,usable-memory = <0x00000000 0x04000000 0x17400000 0x28c00000>;//0x17400000+0x28c00000=0x40000000 1GB DDR
+		aml_reserved_start = <0x06000000>; /**reserved memory start */
+		aml_reserved_end = <0x05000000>;/**reserved memory end */
+		linux,total-memory = <0x40000000>;
 	};
 /// ***************************************************************************************
 ///	-	GIC
@@ -109,17 +109,6 @@ void root_func(){
 		    	6: 318.75M		7: 364.3M					 */
 	};
 	
-/// ***************************************************************************************
-///	-	ION
-//$$ MODULE="ION"
-//$$ DEVICE="ion_dev"
-//$$ L2 PROP_STR = "status"
-    ion_dev{
-        compatible = "amlogic,ion_dev";
-        dev_name = "ion_dev";
-        status = "ok";
-        reg = <0x0bd00000 0x01000000>;//same as following ppmgr
-    };
 /// **************************************************************************************	
 /// -   DISP&MM-FB
 //$$ MODULE = "DISP&MM-FB"	
@@ -133,9 +122,7 @@ void root_func(){
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reg = <0x05100000 0x01000000
-			0x06100000 0x00100000>;
-		rotation = <0>;
+	   	reserve-memory = <0x01000000  0x00100000>;
 		vmode = <1>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P*/
 		display_size_default = <768 1024 768 2048 32>;//1024x768x8=0x600000<0x01000000
 	};
@@ -147,9 +134,8 @@ void root_func(){
 		compatible = "amlogic,mesonfb-ext";
 		dev_name = "mesonfb-ext";
 		status = "ok";
-		reg = <0x06200000 0x01000000
-	       	 0x07200000 0x00100000>;
-		display_size_default = <768 1024 768 2048 32>;//1024x768x8=0x600000<0x01000000
+	   	reserve-memory = <0x01000000  0x00100000>;
+		display_size_default = <768 1024 768 2048 32>;////1024x768x8=0x600000<0x01000000
 	 
 	};
 
@@ -163,8 +149,7 @@ void root_func(){
 		compatible = "amlogic,mesonstream";
 		dev_name = "mesonstream.0";
 		status = "okay";
-		reg = <0x07300000 0x04000000
-			   0x0b300000 0x00a00000>;
+		reserve-memory = <0x04000000 0x00a00000>;
 	};
 
 
@@ -178,7 +163,7 @@ void root_func(){
 		compatible = "amlogic,ppmgr";
 		dev_name = "ppmgr";
 		status = "okay";
-		reg = <0x0bd00000 0x01000000>;		
+		reserve-memory = <0x01000000>;
 	};
 
 /// ***************************************************************************************
@@ -191,7 +176,7 @@ void root_func(){
 		compatible = "amlogic,amvenc_avc";
 		dev_name = "amvenc_avc.0";
 		status = "okay";
-		reg = <0x0cd00000 0x00f00000>;
+		reserve-memory = <0x00f00000>;
 	};
 
 /// ***************************************************************************************
@@ -206,7 +191,7 @@ void root_func(){
         compatible = "amlogic,vdin";
         dev_name = "vdin0";
         status = "ok";
-        reg = <0x0dc00000 0x04000000>;
+	 reserve-memory = <0x04000000>;
         irq = <115>;
         vdin_id = <0>;
 	};
@@ -220,7 +205,7 @@ void root_func(){
         compatible = "amlogic,vdin";
         dev_name = "vdin1";
         status = "ok";
-        reg = <0x11c00000 0x04000000>;
+	 reserve-memory = <0x04000000>;
         irq = <117>;
         vdin_id = <1>;
 	};
@@ -235,10 +220,20 @@ void root_func(){
 		compatible = "amlogic,amlvideo2";
 		dev_name = "amlvideo2.0";
 		status = "okay";
-		reg = <0x15c00000 0x01800000>;
+		reserve-memory = <0x01800000>;
 	};
 
-
+/// ***************************************************************************************
+///	-	ION
+//$$ MODULE="ION"
+//$$ DEVICE="ion_dev"
+//$$ L2 PROP_STR = "status"
+    ion_dev{
+        compatible = "amlogic,ion_dev";
+        dev_name = "ion_dev";
+        status = "ok";
+        reserve-memory = <0x01000000>;
+    };
 /// ***************************************************************************************
 ///	-	DISP&MM-Vout
 //$$ MODULE = "DISP&MM-Vout"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
index 649a37766492..23daa140309d 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
@@ -77,9 +77,9 @@ void root_func(){
 //$$ L2 PROP_U32 6 = "linux,usable-memory"
 	memory{
 		device_type = "memory";
-		aml_reserved_start = <0x05000000>; /**reserved memory start */
-		aml_reserved_end = <0x1e3fffff>;/**reserved memory end */
-		linux,usable-memory = <0x00000000 0x04000000 0x1e400000 0x21c00000>;
+		aml_reserved_start = <0x06000000>; /**reserved memory start */
+		aml_reserved_end = <0x05000000>;/**reserved memory end */
+		linux,total-memory = <0x40000000>;
 	};
 /// ***************************************************************************************
 ///	-	GIC
@@ -110,17 +110,6 @@ void root_func(){
 		    	6: 318.75M		7: 364.3M					 */
 	};
 	
-/// ***************************************************************************************
-///	-	ION
-//$$ MODULE="ION"
-//$$ DEVICE="ion_dev"
-//$$ L2 PROP_STR = "status"
-    ion_dev{
-        compatible = "amlogic,ion_dev";
-        dev_name = "ion_dev";
-        status = "ok";
-        reg = <0x08500000 0x03000000>;
-    };
 /// **************************************************************************************	
 /// -   DISP&MM-FB
 //$$ MODULE = "DISP&MM-FB"	
@@ -133,8 +122,7 @@ void root_func(){
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reg = <0x05100000 0x01000000
-	   	 		0x06100000 0x00100000>;
+	   	reserve-memory = <0x01000000  0x00100000>;
         vmode = <3>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
  		display_size_default = <1920 1080 1920 2160 32>;
 	};
@@ -146,7 +134,7 @@ void root_func(){
 		compatible = "amlogic,deinterlace";
 		dev_name = "deinterlace";
 		status = "ok";
-		reg = <0x06200000 0x02300000>;
+		reserve-memory = <0x02300000>; 	 
 	};
 
 /// ***************************************************************************************
@@ -159,8 +147,7 @@ void root_func(){
 		compatible = "amlogic,mesonstream";
 		dev_name = "mesonstream.0";
 		status = "okay";
-		reg = <0x0c400000 0x10000000
-			   0x1c400000 0x02000000>;
+		reserve-memory = <0x10000000 0x02000000>;
 	};
 
 
@@ -174,7 +161,7 @@ void root_func(){
 		compatible = "amlogic,ppmgr";
 		dev_name = "ppmgr";
 		status = "okay";
-		reg = <0x08500000 0x03000000>;		
+		reserve-memory = <0x03000000>;
 	};
 
 /// ***************************************************************************************
@@ -187,7 +174,7 @@ void root_func(){
 		compatible = "amlogic,amvenc_avc";
 		dev_name = "amvenc_avc.0";
 		status = "okay";
-		reg = <0x0b500000 0x00f00000>;
+		reserve-memory = <0x00f00000>;
 	};
 	
 /// ***************************************************************************************
@@ -215,7 +202,7 @@ void root_func(){
 //        compatible = "amlogic,vdin";
 //        dev_name = "vdin1";
 //        status = "ok";
-//        reg = <0x1e400000 0x04000000>;
+//        reserve-memory = <0x04000000>;
 //        irq = <117>;
 //        vdin_id = <1>;
 //	};
@@ -230,9 +217,20 @@ void root_func(){
 //		compatible = "amlogic,amlvideo2";
 //		dev_name = "amlvideo2.0";
 //		status = "okay";
-//		reg = <0x22400000 0x01800000>;
+//		reserve-memory = <0x01800000>;
 //	};
 
+/// ***************************************************************************************
+///	-	ION
+//$$ MODULE="ION"
+//$$ DEVICE="ion_dev"
+//$$ L2 PROP_STR = "status"
+    ion_dev{
+        compatible = "amlogic,ion_dev";
+        dev_name = "ion_dev";
+        status = "ok";
+        reserve-memory = <0x03000000>;
+    };
 /// ***************************************************************************************
 ///	-	DISP&MM-Vout
 //$$ MODULE = "DISP&MM-Vout"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
index 43d02aa38c13..7a676a3b1e35 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
@@ -76,9 +76,9 @@ void root_func(){
 //$$ L2 PROP_U32 6 = "linux,usable-memory"
 	memory{
 		device_type = "memory";
-		aml_reserved_start = <0x05000000>; /**reserved memory start */
-		aml_reserved_end = <0x1e3fffff>;/**reserved memory end */
-		linux,usable-memory = <0x00000000 0x04000000 0x1e400000 0x61c00000>;
+		aml_reserved_start = <0x06000000>; /**reserved memory start */
+		aml_reserved_end = <0x05000000>;/**reserved memory end */
+		linux,total-memory = <0x80000000>;
 	};
 /// ***************************************************************************************
 ///	-	GIC
@@ -109,17 +109,6 @@ void root_func(){
 		    	6: 318.75M		7: 364.3M					 */
 	};
 	
-/// ***************************************************************************************
-///	-	ION
-//$$ MODULE="ION"
-//$$ DEVICE="ion_dev"
-//$$ L2 PROP_STR = "status"
-    ion_dev{
-        compatible = "amlogic,ion_dev";
-        dev_name = "ion_dev";
-        status = "ok";
-        reg = <0x08500000 0x03000000>;
-    };
 /// **************************************************************************************	
 /// -   DISP&MM-FB
 //$$ MODULE = "DISP&MM-FB"	
@@ -132,8 +121,7 @@ void root_func(){
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reg = <0x05100000 0x01000000
-	   	 		0x06100000 0x00100000>;
+	   	reserve-memory = <0x01000000  0x00100000>;
         vmode = <3>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
  		display_size_default = <1920 1080 1920 2160 32>;
 	};
@@ -145,7 +133,7 @@ void root_func(){
 		compatible = "amlogic,deinterlace";
 		dev_name = "deinterlace";
 		status = "ok";
-		reg = <0x06200000 0x02300000>;
+		reserve-memory = <0x02300000>; 	 
 	};
 
 /// ***************************************************************************************
@@ -158,8 +146,7 @@ void root_func(){
 		compatible = "amlogic,mesonstream";
 		dev_name = "mesonstream.0";
 		status = "okay";
-		reg = <0x0c400000 0x10000000
-			   0x1c400000 0x02000000>;
+		reserve-memory = <0x10000000 0x02000000>;
 	};
 
 
@@ -173,7 +160,7 @@ void root_func(){
 		compatible = "amlogic,ppmgr";
 		dev_name = "ppmgr";
 		status = "okay";
-		reg = <0x08500000 0x03000000>;		
+		reserve-memory = <0x03000000>;
 	};
 
 /// ***************************************************************************************
@@ -186,7 +173,7 @@ void root_func(){
 		compatible = "amlogic,amvenc_avc";
 		dev_name = "amvenc_avc.0";
 		status = "okay";
-		reg = <0x0b500000 0x00f00000>;
+		reserve-memory = <0x00f00000>;
 	};
 	
 /// ***************************************************************************************
@@ -214,7 +201,7 @@ void root_func(){
 //        compatible = "amlogic,vdin";
 //        dev_name = "vdin1";
 //        status = "ok";
-//        reg = <0x1e400000 0x04000000>;
+//        reserve-memory = <0x04000000>;
 //        irq = <117>;
 //        vdin_id = <1>;
 //	};
@@ -229,9 +216,20 @@ void root_func(){
 //		compatible = "amlogic,amlvideo2";
 //		dev_name = "amlvideo2.0";
 //		status = "okay";
-//		reg = <0x22400000 0x01800000>;
+//		reserve-memory = <0x01800000>;
 //	};
 
+/// ***************************************************************************************
+///	-	ION
+//$$ MODULE="ION"
+//$$ DEVICE="ion_dev"
+//$$ L2 PROP_STR = "status"
+    ion_dev{
+        compatible = "amlogic,ion_dev";
+        dev_name = "ion_dev";
+        status = "ok";
+        reserve-memory = <0x03000000>;
+    };
 /// ***************************************************************************************
 ///	-	DISP&MM-Vout
 //$$ MODULE = "DISP&MM-Vout"
diff --git a/arch/arm/boot/dts/amlogic/meson8_skt.dtd b/arch/arm/boot/dts/amlogic/meson8_skt.dtd
index 94f55588d4af..5f4d31e3dbe3 100755
--- a/arch/arm/boot/dts/amlogic/meson8_skt.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_skt.dtd
@@ -76,9 +76,9 @@ void root_func(){
 //$$ L2 PROP_U32 6 = "linux,usable-memory"
 	memory{
 		device_type = "memory";
-		aml_reserved_start = <0x05000000>; /**reserved memory start */
-		aml_reserved_end = <0x27bfffff>;/**reserved memory end */
-		linux,usable-memory = <0x00000000 0x04000000 0x27c00000 0x58400000>;
+		aml_reserved_start = <0x06000000>; /**reserved memory start */
+		aml_reserved_end = <0x05000000>;/**reserved memory end : dtb start for uboot*/
+		linux,total-memory = <0x80000000>;
 	};
 /// ***************************************************************************************
 ///	-	GIC
@@ -109,17 +109,6 @@ void root_func(){
 		    	6: 318.75M		7: 364.3M					 */
 	};
 	
-/// ***************************************************************************************
-///	-	ION
-//$$ MODULE="ION"
-//$$ DEVICE="ion_dev"
-//$$ L2 PROP_STR = "status"
-    ion_dev{
-        compatible = "amlogic,ion_dev";
-        dev_name = "ion_dev";
-        status = "ok";
-        reg = <0x18200000 0x03000000>;
-    };
 /// **************************************************************************************	
 /// -   DISP&MM-FB
 //$$ MODULE = "DISP&MM-FB"	
@@ -132,8 +121,7 @@ void root_func(){
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reg = <0x05100000 0x01000000
-	   	 		0x06100000 0x00100000>;
+	   	reserve-memory = <0x01000000  0x00100000>;
         vmode = <3>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
  		display_size_default = <1920 1080 1920 2160 32>;
 	};
@@ -145,7 +133,8 @@ void root_func(){
 		compatible = "amlogic,deinterlace";
 		dev_name = "deinterlace";
 		status = "ok";
-		reg = <0x25900000 0x02300000>;
+		reserve-memory = <0x02300000>; 	 
+
 	};
 
 /// ***************************************************************************************
@@ -158,8 +147,7 @@ void root_func(){
 		compatible = "amlogic,mesonstream";
 		dev_name = "mesonstream.0";
 		status = "okay";
-		reg = <0x06200000 0x10000000
-			   0x16200000 0x02000000>;
+		reserve-memory = <0x10000000 0x02000000>;
 	};
 
 
@@ -173,7 +161,7 @@ void root_func(){
 		compatible = "amlogic,ppmgr";
 		dev_name = "ppmgr";
 		status = "okay";
-		reg = <0x18200000 0x03000000>;		
+		reserve-memory = <0x03000000>;
 	};
 
 /// ***************************************************************************************
@@ -186,7 +174,7 @@ void root_func(){
 		compatible = "amlogic,amvenc_avc";
 		dev_name = "amvenc_avc.0";
 		status = "okay";
-		reg = <0x1b200000 0x00f00000>;
+		reserve-memory = <0x00f00000>;
 	};
 
 /// ***************************************************************************************
@@ -214,7 +202,7 @@ void root_func(){
         compatible = "amlogic,vdin";
         dev_name = "vdin0";
         status = "ok";
-        reg = <0x1c100000 0x04000000>;
+	 reserve-memory = <0x04000000>;
         irq = <115>;
         vdin_id = <0>;
 	};
@@ -228,7 +216,7 @@ void root_func(){
         compatible = "amlogic,vdin";
         dev_name = "vdin1";
         status = "ok";
-        reg = <0x20100000 0x04000000>;
+	 reserve-memory = <0x04000000>;
         irq = <117>;
         vdin_id = <1>;
 	};
@@ -243,10 +231,20 @@ void root_func(){
 		compatible = "amlogic,amlvideo2";
 		dev_name = "amlvideo2.0";
 		status = "okay";
-		reg = <0x24100000 0x01800000>;
+		reserve-memory = <0x01800000>;
 	};
 
-
+/// ***************************************************************************************
+///	-	ION
+//$$ MODULE="ION"
+//$$ DEVICE="ion_dev"
+//$$ L2 PROP_STR = "status"
+    ion_dev{
+        compatible = "amlogic,ion_dev";
+        dev_name = "ion_dev";
+        status = "ok";
+        reserve-memory = <0x03000000>;
+    };
 /// ***************************************************************************************
 ///	-	DISP&MM-Vout
 //$$ MODULE = "DISP&MM-Vout"
diff --git a/arch/arm/kernel/devtree.c b/arch/arm/kernel/devtree.c
index b794a31e6f71..b1c7a40eefa6 100755
--- a/arch/arm/kernel/devtree.c
+++ b/arch/arm/kernel/devtree.c
@@ -243,6 +243,10 @@ struct machine_desc * __init setup_machine_fdt(unsigned int dt_phys)
 	of_scan_flat_dt(early_init_dt_scan_chosen, boot_command_line);
 	/* Initialize {size,address}-cells info */
 	of_scan_flat_dt(early_init_dt_scan_root, NULL);
+#if defined(CONFIG_PLAT_MESON)
+	init_reserve_mgr();
+	of_scan_flat_dt(early_init_dt_scan_reserve_memory, NULL);
+#endif
 	/* Setup memory, calling early_init_dt_add_memory_arch */
 	of_scan_flat_dt(early_init_dt_scan_memory, NULL);
 
diff --git a/drivers/amlogic/amports/amstream.c b/drivers/amlogic/amports/amstream.c
index cba4db3a1c01..48bd5e4277d9 100755
--- a/drivers/amlogic/amports/amstream.c
+++ b/drivers/amlogic/amports/amstream.c
@@ -65,6 +65,7 @@
 #include "ampotrs_priv.h"
 
 #include <linux/of.h>
+#include <linux/of_fdt.h>
 
 #define DEVICE_NAME "amstream-dev"
 #define DRIVER_NAME "amstream"
@@ -1846,6 +1847,7 @@ static struct class amstream_class = {
         .class_attrs = amstream_class_attrs,
     };
 
+static struct resource memobj;
 static int  amstream_probe(struct platform_device *pdev)
 {
     int i;
@@ -1877,9 +1879,21 @@ static int  amstream_probe(struct platform_device *pdev)
 
         goto error2;
     }
-
+#if 0
     vdec_set_resource(platform_get_resource(pdev, IORESOURCE_MEM, 0), (void *)&amstream_dec_info);
-
+#else
+    res = &memobj;
+    r = find_reserve_block(pdev->dev.of_node->name,0);
+    if(r < 0){
+        printk("can not find %s%d reserve block\n",pdev->dev.of_node->name,0);
+	 r = -EFAULT;
+	 goto error2;
+    }
+    res->start = (phys_addr_t)get_reserve_block_addr(r);
+    res->end = res->start+ (phys_addr_t)get_reserve_block_size(r)-1;
+    res->flags = IORESOURCE_MEM;
+    vdec_set_resource(res, (void *)&amstream_dec_info);
+#endif
     amstream_dev_class = class_create(THIS_MODULE, DEVICE_NAME);
 
     for (st = &ports[0], i = 0; i < MAX_AMSTREAM_PORT_NUM; i++, st++) {
@@ -1895,7 +1909,19 @@ static int  amstream_probe(struct platform_device *pdev)
         goto error3;
     }
 
+#if 0
     res = platform_get_resource(pdev, IORESOURCE_MEM, 1);
+#else
+    res = &memobj;
+    r = find_reserve_block(pdev->dev.of_node->name,1);
+    if(r < 0){
+        printk("can not find %s%d reserve block\n",pdev->dev.of_node->name,1);
+	 r = -EFAULT;
+	 goto error3;
+    }
+    res->start = (phys_addr_t)get_reserve_block_addr(r);
+    res->end = res->start+ (phys_addr_t)get_reserve_block_size(r)-1;
+#endif
     if (!res) {
         printk("Can not obtain I/O memory, and will allocate stream buffer!\n");
 
diff --git a/drivers/amlogic/amports/encoder.c b/drivers/amlogic/amports/encoder.c
index e8ce21759037..88b0b26c85dd 100755
--- a/drivers/amlogic/amports/encoder.c
+++ b/drivers/amlogic/amports/encoder.c
@@ -28,6 +28,8 @@
 #include <linux/amlogic/amports/vframe_receiver.h>
 #include "vdec_reg.h"
 #include <linux/delay.h>
+#include <linux/of.h>
+#include <linux/of_fdt.h>
 
 #define ENC_CANVAS_OFFSET  AMVENC_CANVAS_INDEX
 
@@ -1459,17 +1461,29 @@ int uninit_avc_device(void)
     return 0;
 }
 
+static struct resource memobj;
 static int amvenc_avc_probe(struct platform_device *pdev)
 {
     struct resource *mem;
+    int idx;
 
     amlog_level(LOG_LEVEL_INFO, "amvenc_avc probe start.\n");
 
+#if 0
     if (!(mem = platform_get_resource(pdev, IORESOURCE_MEM, 0))) {
         amlog_level(LOG_LEVEL_ERROR, "amvenc_avc memory resource undefined.\n");
         return -EFAULT;
     }
-
+#else
+    mem = &memobj;
+    idx = find_reserve_block(pdev->dev.of_node->name,0);
+    if(idx < 0){
+	 amlog_level(LOG_LEVEL_ERROR, "amvenc_avc memory resource undefined.\n");
+        return -EFAULT;
+    }
+    mem->start = (phys_addr_t)get_reserve_block_addr(idx);
+    mem->end = mem->start+ (phys_addr_t)get_reserve_block_size(idx)-1;
+#endif
     gAmvencbuff.buf_start = mem->start;
     gAmvencbuff.buf_size = mem->end - mem->start + 1;
 
diff --git a/drivers/amlogic/deinterlace/deinterlace.c b/drivers/amlogic/deinterlace/deinterlace.c
index 97b035db8640..b28e4335ba0a 100755
--- a/drivers/amlogic/deinterlace/deinterlace.c
+++ b/drivers/amlogic/deinterlace/deinterlace.c
@@ -37,7 +37,7 @@
 #include <linux/cdev.h>
 #include <linux/proc_fs.h>
 #include <linux/list.h>
-
+#include <linux/of_fdt.h>
 //#include <linux/aml_common.h>
 #include <asm/uaccess.h>
 #include <mach/am_regs.h>
@@ -6301,6 +6301,7 @@ const static struct file_operations di_fops = {
 #endif
 };
 
+static struct resource memobj;
 static int di_probe(struct platform_device *pdev)
 {
     int r, i;
@@ -6372,11 +6373,22 @@ static int di_probe(struct platform_device *pdev)
     device_create_file(di_device.dev, &dev_attr_status);
     device_create_file(di_device.dev, &dev_attr_provider_vframe_status);
 
+#if 0
     if (!(mem = platform_get_resource(pdev, IORESOURCE_MEM, 0)))
     {
     	  pr_error("\ndeinterlace memory resource undefined.\n");
         return -EFAULT;
     }
+#else
+    mem = &memobj;
+    r = find_reserve_block(pdev->dev.of_node->name,0);
+    if(r < 0){
+        pr_error("\ndeinterlace memory resource undefined.\n");
+        return -EFAULT;
+    }
+    mem->start = (phys_addr_t)get_reserve_block_addr(r);
+    mem->end = mem->start+ (phys_addr_t)get_reserve_block_size(r)-1;
+#endif
 		for(i=0; i<USED_LOCAL_BUF_MAX; i++){
     	used_local_buf_index[i] = -1;
  		}
diff --git a/drivers/amlogic/display/aml_logo/logo_display/logo_output.c b/drivers/amlogic/display/aml_logo/logo_display/logo_output.c
index d161595ca61f..33fcbfec281b 100755
--- a/drivers/amlogic/display/aml_logo/logo_display/logo_output.c
+++ b/drivers/amlogic/display/aml_logo/logo_display/logo_output.c
@@ -12,6 +12,8 @@
 #include  "logo.h"
 #include	"amlogo_log.h"
 #include <linux/amlogic/amlog.h>
+#include <linux/of.h>
+#include <linux/of_fdt.h>
 static  LIST_HEAD(output_dev_line);
 static  output_dev_list_t aml_output_dev[LOGO_DEV_MAX];
 
@@ -23,7 +25,8 @@ static  output_dev_list_t aml_output_dev[LOGO_DEV_MAX];
 	if (strncmp(name, dev_name(dev),DEVICE_NAME_LEN) == 0)
 		return 1;
 	return 0;
-} 
+}
+ static struct resource memobj;
  int   setup_logo_platform_resource(logo_object_t *logo)
 {
 	int  i;
@@ -31,6 +34,7 @@ static  output_dev_list_t aml_output_dev[LOGO_DEV_MAX];
 	struct  device  *dev=NULL;
 	struct platform_device  * platform_dev;
 	struct resource * res; 
+	int idx;
 
 	for(i=0;i<LOGO_DEV_MEM;i++)
 	{
@@ -44,7 +48,18 @@ static  output_dev_list_t aml_output_dev[LOGO_DEV_MAX];
 			
 			platform_dev =dev_to_platformdev(dev) ;
 			amlog_mask_level(LOG_MASK_DEVICE,LOG_LEVEL_LOW,"got platform resource\n");
+#if 0			
 			res=platform_get_resource(platform_dev,IORESOURCE_MEM,num);//something special.
+#else
+			res = &memobj;
+			idx = find_reserve_block(platform_dev->dev.of_node->name,num);
+			if(idx < 0){
+				amlog_mask_level(LOG_LEVEL_HIGH,"can not find %s%d reserve block\n",platform_dev->dev.of_node->name,num);
+				continue;
+			}
+			res->start = (phys_addr_t)get_reserve_block_addr(idx);
+			res->end = res->start+ (phys_addr_t)get_reserve_block_size(idx)-1;
+#endif
 			if(res)
 			{
 				amlog_mask_level(LOG_MASK_DEVICE,LOG_LEVEL_LOW,"resource: start=0x%x,end=0x%x\r\n",res->start,res->end);
diff --git a/drivers/amlogic/display/osd/osd_main.c b/drivers/amlogic/display/osd/osd_main.c
index 372740637135..118a98219b84 100755
--- a/drivers/amlogic/display/osd/osd_main.c
+++ b/drivers/amlogic/display/osd/osd_main.c
@@ -46,6 +46,7 @@
 #include <linux/slab.h>
 #include <asm/uaccess.h>
 #include <linux/of.h>
+#include <linux/of_fdt.h>
 #include "osd_log.h"
 #include <linux/amlogic/amlog.h>
 #ifdef CONFIG_HAS_EARLYSUSPEND
@@ -1478,6 +1479,7 @@ void osd_resume_early(void)
 EXPORT_SYMBOL(osd_resume_early);
 #endif
 
+static struct resource memobj;
 static int 
 osd_probe(struct platform_device *pdev)
 {
@@ -1533,6 +1535,7 @@ osd_probe(struct platform_device *pdev)
 	vinfo = get_current_vinfo();
     	for (index=0;index<OSD_COUNT;index++)
     	{
+#if 0
     		//platform resource 
 		if (!(mem = platform_get_resource(pdev, IORESOURCE_MEM, index)))
 		{
@@ -1547,6 +1550,17 @@ osd_probe(struct platform_device *pdev)
 		{
 			continue ;
 		}
+#else
+		mem = &memobj;
+		ret = find_reserve_block(pdev->dev.of_node->name,index);
+		if(ret < 0){
+			amlog_level(LOG_LEVEL_HIGH,"can not find %s%d reserve block\n",pdev->dev.of_node->name,index);
+			r = -EFAULT;
+			goto failed2;
+		}
+		mem->start = (phys_addr_t)get_reserve_block_addr(ret);
+		mem->end = mem->start+ (phys_addr_t)get_reserve_block_size(ret)-1;
+#endif
 		fbi = framebuffer_alloc(sizeof(struct myfb_dev), &pdev->dev);
     		if(!fbi)
     		{
diff --git a/drivers/amlogic/display/osd_ext/osd_main.c b/drivers/amlogic/display/osd_ext/osd_main.c
index 7216b6c9aab5..1abce9a9229b 100755
--- a/drivers/amlogic/display/osd_ext/osd_main.c
+++ b/drivers/amlogic/display/osd_ext/osd_main.c
@@ -40,6 +40,7 @@
 #include <linux/sysfs.h>
 #include <linux/file.h>
 #include <linux/fdtable.h>
+#include <linux/of_fdt.h>
 #include <linux/console.h>
 #include <linux/slab.h>
 #include <asm/uaccess.h>
@@ -1348,6 +1349,7 @@ void osd_ext_resume_early(void)
 EXPORT_SYMBOL(osd_ext_resume_early);
 #endif
 
+static struct resource memobj;
 static int 
 osd_ext_probe(struct platform_device *pdev)
 {
@@ -1372,6 +1374,7 @@ osd_ext_probe(struct platform_device *pdev)
 	vinfo = get_current_vinfo2();
 	for (index = 0; index < OSD_COUNT; index++) {
 		//platform resource
+#if 0
 		if (!(mem = platform_get_resource(pdev, IORESOURCE_MEM, index))) {
 			amlog_level(LOG_LEVEL_HIGH, "No frame buffer memory define.\n");
 			r = -EFAULT;
@@ -1383,7 +1386,17 @@ osd_ext_probe(struct platform_device *pdev)
 		if (!mem || mem->start == 0 || mem->end == 0 || mem->start == mem->end) {
 			continue ;
 		}
-
+#else
+		mem = &memobj;
+		ret = find_reserve_block(pdev->dev.of_node->name,index);
+		if(ret < 0){
+			amlog_level(LOG_LEVEL_HIGH,"can not find %s%d reserve block\n",pdev->dev.of_node->name,index);
+			r = -EFAULT;
+			goto failed2;
+		}
+		mem->start = (phys_addr_t)get_reserve_block_addr(ret);
+		mem->end = mem->start+ (phys_addr_t)get_reserve_block_size(ret)-1;
+#endif
 		fbi = framebuffer_alloc(sizeof(struct myfb_dev), &pdev->dev);
 		if (!fbi) {
 			r = -ENOMEM;
diff --git a/drivers/amlogic/ion_dev/dev_ion.c b/drivers/amlogic/ion_dev/dev_ion.c
index 78e7cac06293..1e95b3215ebc 100644
--- a/drivers/amlogic/ion_dev/dev_ion.c
+++ b/drivers/amlogic/ion_dev/dev_ion.c
@@ -19,6 +19,8 @@
 #include <linux/slab.h>
 #include <linux/module.h>
 #include <ion_priv.h>
+#include <linux/of.h>
+#include <linux/of_fdt.h>
 
 MODULE_DESCRIPTION("AMLOGIC ION driver");
 MODULE_LICENSE("GPL");
@@ -40,6 +42,7 @@ static int num_heaps;
 static struct ion_heap **heaps;
 static struct ion_platform_heap my_ion_heap[MAX_HEAP];
 
+static struct resource memobj;
 int dev_ion_probe(struct platform_device *pdev) {
     int err;
     int i;
@@ -50,7 +53,18 @@ int dev_ion_probe(struct platform_device *pdev) {
     my_ion_heap[0].type = ION_HEAP_TYPE_SYSTEM;
     my_ion_heap[0].id = ION_HEAP_TYPE_SYSTEM;
     my_ion_heap[0].name = "vmalloc_ion";
+#if 0
     res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+#else
+    res = &memobj;
+    i = find_reserve_block(pdev->dev.of_node->name,0);
+    if(i < 0){
+        printk("\ndev_ion memory resource undefined.\n");
+        return -EFAULT;
+    }
+    res->start = (phys_addr_t)get_reserve_block_addr(i);
+    res->end = res->start+ (phys_addr_t)get_reserve_block_size(i)-1;
+#endif
     if (res) {
         num_heaps = 2;
         my_ion_heap[1].type = ION_HEAP_TYPE_CARVEOUT;//ION_HEAP_TYPE_CHUNK;//ION_HEAP_TYPE_CARVEOUT;
diff --git a/drivers/amlogic/ppmgr/ppmgr_drv.c b/drivers/amlogic/ppmgr/ppmgr_drv.c
index 231fc7aef9d3..28e664107c35 100755
--- a/drivers/amlogic/ppmgr/ppmgr_drv.c
+++ b/drivers/amlogic/ppmgr/ppmgr_drv.c
@@ -19,7 +19,7 @@
 #include <linux/amlogic/amports/vframe.h>
 #include <linux/amlogic/amports/vframe_provider.h>
 #include <linux/amlogic/amports/vframe_receiver.h>
-
+#include <linux/of_fdt.h>
 
 
 #include "ppmgr_log.h"
@@ -1173,20 +1173,31 @@ int uninit_ppmgr_device(void)
 MODULE_AMLOG(AMLOG_DEFAULT_LEVEL, 0xff, LOG_LEVEL_DESC, LOG_MASK_DESC);
 
 static struct platform_device *ppmgr_dev0 = NULL;
-
+static struct resource memobj;
 /* for driver. */
 static int ppmgr_driver_probe(struct platform_device *pdev)
 {
     char* buf_start;
     unsigned int buf_size;
     struct resource *mem;
+    int idx;
 
+#if 0
     if (!(mem = platform_get_resource(pdev, IORESOURCE_MEM, 0)))
     {
         amlog_level(LOG_LEVEL_HIGH, "ppmgr memory resource undefined.\n");
         return -EFAULT;
     }
-
+#else
+    mem = &memobj;
+    idx = find_reserve_block(pdev->dev.of_node->name,0);
+    if(idx < 0){
+	 amlog_level(LOG_LEVEL_HIGH, "ppmgr memory resource undefined.\n");
+        return -EFAULT;
+    }
+    mem->start = (phys_addr_t)get_reserve_block_addr(idx);
+    mem->end = mem->start+ (phys_addr_t)get_reserve_block_size(idx)-1;
+#endif
     buf_start = (char *)mem->start;
     buf_size = mem->end - mem->start + 1;
     set_ppmgr_buf_info((char *)mem->start,buf_size);
diff --git a/drivers/amlogic/tvin/vdin/vdin_drv.c b/drivers/amlogic/tvin/vdin/vdin_drv.c
index 67414deb7526..4bc094e3ebb7 100755
--- a/drivers/amlogic/tvin/vdin/vdin_drv.c
+++ b/drivers/amlogic/tvin/vdin/vdin_drv.c
@@ -34,6 +34,7 @@
 #include <asm/fiq.h>
 #include <asm/div64.h>
 #include <linux/of.h>
+#include <linux/of_fdt.h>
 /* Amlogic Headers */
 #include <linux/amlogic/amports/canvas.h>
 #include <mach/am_regs.h>
@@ -2634,6 +2635,7 @@ static void vdin_delete_device(int minor)
 	device_destroy(vdin_clsp, devno);
 }
 
+static struct resource memobj;
 static int vdin_drv_probe(struct platform_device *pdev)
 {
 	int ret = 0;
@@ -2689,7 +2691,18 @@ static int vdin_drv_probe(struct platform_device *pdev)
 		goto fail_create_dev_file;
 	}
 	/* get memory address from resource */
+#if 0	
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+#else
+    res = &memobj;
+    ret = find_reserve_block(pdev->dev.of_node->name,0);
+    if(ret < 0){
+        pr_err("\nvdin memory resource undefined.\n");
+        return -EFAULT;
+    }
+    res->start = (phys_addr_t)get_reserve_block_addr(ret);
+    res->end = res->start+ (phys_addr_t)get_reserve_block_size(ret)-1;
+#endif
 	if (!res) {
 		pr_err("%s: can't get mem resource\n", __func__);
 		ret = -ENXIO;
diff --git a/drivers/amlogic/video_dev/amlvideo2.c b/drivers/amlogic/video_dev/amlvideo2.c
index d84ad86cb9bb..ad4a52d52c35 100755
--- a/drivers/amlogic/video_dev/amlvideo2.c
+++ b/drivers/amlogic/video_dev/amlvideo2.c
@@ -48,6 +48,7 @@
 #endif
 #ifdef CONFIG_USE_OF
 #include <linux/of.h>
+#include <linux/of_fdt.h>
 #endif
 
 #define AVMLVIDEO2_MODULE_NAME "amlvideo2"
@@ -1924,6 +1925,7 @@ static int amlvideo2_release_node(struct amlvideo2_device *vid_dev)
 	return 0;
 }
 
+static struct resource memobj;
 static int amlvideo2_create_node(struct platform_device *pdev)
 {
 	int ret = 0, i = 0, j = 0;
@@ -1940,8 +1942,19 @@ static int amlvideo2_create_node(struct platform_device *pdev)
 
 	for(i = 0; i<vid_dev->node_num;i++){
 		vid_dev->node[i] = NULL;
-		ret = -ENOMEM;		
+		ret = -ENOMEM;	
+#if 0		
 		res = platform_get_resource(pdev, IORESOURCE_MEM, i);
+#else
+		res = &memobj;
+    		ret = find_reserve_block(pdev->dev.of_node->name,i);
+    		if(ret < 0){
+        		printk("\namlvideo2 memory resource undefined.\n");
+        		return -EFAULT;
+    		}
+    		res->start = (phys_addr_t)get_reserve_block_addr(ret);
+    		res->end = res->start+ (phys_addr_t)get_reserve_block_size(ret)-1;
+#endif
 		if(!res)
 			break;
 
diff --git a/drivers/of/fdt.c b/drivers/of/fdt.c
index 9c4274f7ff0a..8e577724b075 100755
--- a/drivers/of/fdt.c
+++ b/drivers/of/fdt.c
@@ -370,7 +370,7 @@ static void __unflatten_device_tree(struct boot_param_header *blob,
 		return;
 	}
 
-	pr_debug("Unflattening device tree:\n");
+	pr_debug("Unflattening device tree:%x\n",(unsigned int)blob);
 	pr_debug("magic: %08x\n", be32_to_cpu(blob->magic));
 	pr_debug("size: %08x\n", be32_to_cpu(blob->totalsize));
 	pr_debug("version: %08x\n", be32_to_cpu(blob->version));
@@ -615,6 +615,110 @@ u64 __init dt_mem_next_cell(int s, __be32 **cellp)
 #if defined(CONFIG_PLAT_MESON)
 extern unsigned long long aml_reserved_start;
 extern unsigned long long aml_reserved_end;
+
+#define MAX_RESERVE_BLOCK  32			
+//limit: reserve block < 32
+#define DSP_MEM_SIZE	0x100000
+#define MEM_BLOCK1_START	0
+#define MEM_BLOCK1_SIZE	0x4000000
+
+struct reserve_mem{
+	unsigned long long startaddr;
+	unsigned long long size;
+	char name[16];					//limit: device name must < 14;
+};
+
+struct reserve_mgr{
+	int count;
+	unsigned long long current_addr;
+	struct reserve_mem reserve[MAX_RESERVE_BLOCK];
+};
+
+struct reserve_mgr Reserve_Manager;
+struct reserve_mgr * pReserve_Manager;
+
+
+int init_reserve_mgr(void)
+{
+	pReserve_Manager = &Reserve_Manager;
+	pReserve_Manager->count = 0;
+	pReserve_Manager->current_addr=0;
+}
+
+unsigned long long get_reserve_end(void)
+{
+	return pReserve_Manager->current_addr+aml_reserved_start+DSP_MEM_SIZE-1;
+}
+
+int find_reserve_block(char * name,int idx)
+{
+	int i;
+
+	for(i=0;i<pReserve_Manager->count;i++)
+	{
+		if((strncmp(pReserve_Manager->reserve[i].name,name,strlen(name))==0)&&(pReserve_Manager->reserve[i].name[strlen(name)]=='0'+idx))
+			return i;
+	}
+
+	return -1;
+}
+
+unsigned long long get_reserve_block_addr(int blockid)
+{
+	if(blockid >= MAX_RESERVE_BLOCK)
+		printk("error: reserve block count is larger than MAX_RESERVE_BLOCK,please reset the code\n");
+	return pReserve_Manager->reserve[blockid].startaddr+aml_reserved_start+DSP_MEM_SIZE;
+}
+
+unsigned long long get_reserve_block_size(int blockid)
+{
+	if(blockid >= MAX_RESERVE_BLOCK)
+		printk("error: reserve block count is larger than MAX_RESERVE_BLOCK,please reset the code\n");
+	return pReserve_Manager->reserve[blockid].size;
+}
+
+
+/**
+ * early_init_dt_scan_memory - Look for an parse memory nodes
+ */
+int __init early_init_dt_scan_reserve_memory(unsigned long node, const char *uname,
+				     int depth, void *data)
+{
+	__be32 *mem, *endp;
+	unsigned long l;
+	int idx=0;
+	mem = of_get_flat_dt_prop(node, "reserve-memory", &l);
+	if (mem == NULL)
+		return 0;
+	
+	endp = mem + (l / sizeof(__be32));
+	while ((endp - mem) >= dt_root_size_cells) {
+		u64 size;
+
+		size = dt_mem_next_cell(dt_root_size_cells, &mem);
+
+		if (size == 0)
+			continue;
+
+		pReserve_Manager->reserve[pReserve_Manager->count].startaddr = pReserve_Manager->current_addr;
+		pReserve_Manager->reserve[pReserve_Manager->count].size = size;
+		strcpy(pReserve_Manager->reserve[pReserve_Manager->count].name,uname);
+		pReserve_Manager->reserve[pReserve_Manager->count].name[strlen(uname)] = '0'+idx;
+		pReserve_Manager->reserve[pReserve_Manager->count].name[strlen(uname)+1] = 0;
+
+		pr_info("name=%s,startaddr = %llx,size=%llx\n",pReserve_Manager->reserve[pReserve_Manager->count].name,pReserve_Manager->reserve[pReserve_Manager->count].startaddr,pReserve_Manager->reserve[pReserve_Manager->count].size);
+		pReserve_Manager->current_addr +=size;
+		pReserve_Manager->count +=1;
+		idx++;
+
+		if(pReserve_Manager->count >= MAX_RESERVE_BLOCK){
+			printk("error: reserve block count is larger than MAX_RESERVE_BLOCK,please reset the code\n");
+			break;
+		}
+	}
+
+	return 0;
+}
 #endif
 /**
  * early_init_dt_scan_memory - Look for an parse memory nodes
@@ -643,15 +747,29 @@ int __init early_init_dt_scan_memory(unsigned long node, const char *uname,
 		printk("error: can not get reserved mem start for AML\n");
 	else
 		aml_reserved_start = of_read_number(reg,1);
-	pr_debug("reserved_start is %llx \n ",aml_reserved_start);
+	pr_info("reserved_start is %llx \n ",aml_reserved_start);
+
 	reg = of_get_flat_dt_prop(node, "aml_reserved_end", &l);
 	if (reg == NULL)
 		printk("error: can not get reserved mem end for AML\n");
 	else
-		aml_reserved_end =  of_read_number(reg,1);
-	pr_debug("reserved_end is %llx \n ",aml_reserved_end);
-#endif
-	
+		aml_reserved_end = of_read_number(reg,1);
+	early_init_dt_add_memory_arch(MEM_BLOCK1_START,MEM_BLOCK1_SIZE);
+	early_init_dt_add_memory_arch(aml_reserved_end,aml_reserved_start-aml_reserved_end);
+
+	aml_reserved_end = get_reserve_end();
+	pr_info("reserved_end is %llx \n ",aml_reserved_end);
+
+	u64 total;
+	reg = of_get_flat_dt_prop(node, "linux,total-memory", &l);
+	if (reg == NULL)
+		printk("error: can not get total-memory for AML\n");
+	else
+		total =  of_read_number(reg,1);
+
+	early_init_dt_add_memory_arch(aml_reserved_end+1,total-aml_reserved_end-1);
+	pr_info("total is %llx \n ",total);
+#else
 	reg = of_get_flat_dt_prop(node, "linux,usable-memory", &l);
 	if (reg == NULL)
 		reg = of_get_flat_dt_prop(node, "reg", &l);
@@ -677,6 +795,7 @@ int __init early_init_dt_scan_memory(unsigned long node, const char *uname,
 		early_init_dt_add_memory_arch(base, size);
 	}
 
+#endif
 	return 0;
 }
 
diff --git a/include/linux/amlogic/amports/dsp_register.h b/include/linux/amlogic/amports/dsp_register.h
index e9afbf9c1500..bb223bf908b4 100755
--- a/include/linux/amlogic/amports/dsp_register.h
+++ b/include/linux/amlogic/amports/dsp_register.h
@@ -17,9 +17,9 @@ dsp_register.h
 
 #define AUDIO_DSP_MEM_SIZE		 S_1M
 #ifdef CONFIG_ARCH_MESON8
-#define AUDIO_DSP_START_PHY_ADDR 0x05000000
+#define AUDIO_DSP_START_PHY_ADDR 0x06000000
 #else
-#define AUDIO_DSP_START_PHY_ADDR 0x84000000
+#define AUDIO_DSP_START_PHY_ADDR 0x86000000
 #endif
 #define AUDIO_DSP_START_ADDR	((unsigned)phys_to_virt(AUDIO_DSP_START_PHY_ADDR))//((SYS_MEM_START+SYS_MEM_SIZE)-AUDIO_DSP_MEM_SIZE)
 #define AUDIO_DSP_END_ADDR		((AUDIO_DSP_START_ADDR+AUDIO_DSP_MEM_SIZE))
diff --git a/include/linux/of_fdt.h b/include/linux/of_fdt.h
index ed136ad698ce..d0d231ae246f 100644
--- a/include/linux/of_fdt.h
+++ b/include/linux/of_fdt.h
@@ -96,6 +96,14 @@ extern int early_init_dt_scan_chosen(unsigned long node, const char *uname,
 extern void early_init_dt_check_for_initrd(unsigned long node);
 extern int early_init_dt_scan_memory(unsigned long node, const char *uname,
 				     int depth, void *data);
+#if defined(CONFIG_PLAT_MESON)
+extern int early_init_dt_scan_reserve_memory(unsigned long node, const char *uname,
+				     int depth, void *data);
+extern int init_reserve_mgr(void);
+extern int find_reserve_block(char * name,int idx);
+extern unsigned long long get_reserve_block_addr(int blockid);
+extern unsigned long long get_reserve_block_size(int blockid);
+#endif
 extern void early_init_dt_add_memory_arch(u64 base, u64 size);
 extern void * early_init_dt_alloc_memory_arch(u64 size, u64 align);
 extern u64 dt_mem_next_cell(int s, __be32 **cellp);
-- 
2.19.0

